internal/controller/management/serviceaccount/serviceaccounts.go

```diff
@@ -0,0 +1,229 @@
+package upgrade
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	kubeerr "k8s.io/apimachinery/pkg/api/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/internal/logging"
+)
+
+const (
+	ComponentLabelKey = "app.kubernetes.io/component"
+	ControllerLabelValue = "controller"
+)
+
+type ReconcilerConfig struct {
+	KargoNamespace string `envconfig:"KARGO_NAMESPACE" required:"true"`
+}
+
+// ServiceAccountReconciler Reconciles for ServiceAccounts
+type ServiceAccountReconciler struct {
+	cfg ReconcilerConfig
+	client.Client
+}
+
+// SetupReconcilerWithManager initializes a reconciler
+// and registers that reconciler with the provided Manager.
+func SetupReconcilerWithManager(kargoMgr manager.Manager) error {
+	return ctrl.NewControllerManagedBy(kargoMgr).
+		For(&corev1.ServiceAccount{}).
+		WithEventFilter(predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool {
+				return hasControllerLabel(e.Object.(*corev1.ServiceAccount))
```

Comment:
Initial syncs of existing Stages on controller restart are treated as create events, so this filter shouldn't be here or else this could happen:

1. ServiceAccount has appropriate label, receives corresponding permissions.
2. Controller shuts down
3. ServiceAccount has label removed
4. Controller starts up
5. This filter prevents removal of the permissions.